examples/goproxy-socks/socksclient.go

```diff
@@ -0,0 +1,253 @@
+			}
+		} else {
+			if len(host) > 255 {
+				return nil, errors.New("FQDN too long")
+			}
+			b = append(b, AddrTypeFQDN)
+			b = append(b, byte(len(host)))
+			b = append(b, host...)
+		}
+		b = append(b, byte(port>>8), byte(port))
+		if _, err = c.Write(b); err != nil {
+			return nil, err
+		}
+
+		if _, err = io.ReadFull(c, b[:4]); err != nil {
+			return nil, err
+		}
+		if b[0] != Version5 {
+			return nil, errors.New("unexpected protocol version " + strconv.Itoa(int(b[0])))
+		}
+		if cmdErr := Reply(b[1]); cmdErr != StatusSucceeded {
+			return nil, errors.New("unknown error " + cmdErr.String())
+		}
+		if b[2] != 0 {
+			return nil, errors.New("non-zero reserved field")
+		}
+		l := 2
+		var a Addr
+		switch b[3] {
+		case AddrTypeIPv4:
+			l += net.IPv4len
+			a.IP = make(net.IP, net.IPv4len)
+		case AddrTypeIPv6:
+			l += net.IPv6len
+			a.IP = make(net.IP, net.IPv6len)
+		case AddrTypeFQDN:
+			if _, err := io.ReadFull(c, b[:1]); err != nil {
+				return nil, err
+			}
+			l += int(b[0])
+		default:
+			return nil, errors.New("unknown address type " + strconv.Itoa(int(b[3])))
+		}
+		if cap(b) < l {
+			b = make([]byte, l)
+		} else {
+			b = b[:l]
+		}
+		if _, err = io.ReadFull(c, b); err != nil {
+			return nil, err
+		}
+		if a.IP != nil {
+			copy(a.IP, b)
+		} else {
+			a.Name = string(b[:len(b)-2])
+		}
+		a.Port = int(b[len(b)-2])<<8 | int(b[len(b)-1])
+		return c, nil
+	}, nil
+}
+
+const (
+	authUsernamePasswordVersion = 0x01
+	authStatusSucceeded         = 0x00
+)
+
+func (up *SocksAuth) Authenticate(ctx context.Context, rw io.ReadWriter, auth AuthMethod) error {
+	switch auth {
+	case AuthMethodNotRequired:
+		return nil
+	case AuthMethodUsernamePassword:
+		if len(up.Username) == 0 || len(up.Username) > 255 || len(up.Password) > 255 {
+			return errors.New("invalid username/password")
+		}
+		b := []byte{authUsernamePasswordVersion}
+		b = append(b, byte(len(up.Username)))
+		b = append(b, up.Username...)
+		b = append(b, byte(len(up.Password)))
+		b = append(b, up.Password...)
+
+		if _, err := rw.Write(b); err != nil {
+			return err
+		}
+		if _, err := io.ReadFull(rw, b[:2]); err != nil {
+			return err
+		}
+		if b[0] != authUsernamePasswordVersion {
+			return errors.New("invalid username/password version")
+		}
+		if b[1] != authStatusSucceeded {
+			return errors.New("username/password authentication failed")
+		}
+		return nil
+	}
+	return errors.New("unsupported authentication method " + strconv.Itoa(int(auth)))
+}
+
+func splitHostPort(address string) (string, int, error) {
+	host, port, err := net.SplitHostPort(address)
+	if err != nil {
+		return "", 0, err
+	}
+	portnum, err := strconv.Atoi(port)
+	if err != nil {
+		return "", 0, err
+	}
+	if 1 > portnum || portnum > 0xffff {
+		return "", 0, errors.New("port number out of range " + port)
+	}
+	ip, err := net.ResolveIPAddr("ip", host)
+	if err != nil {
+		return "", 0, err
+	}
+	return ip.String(), portnum, nil
+}
```

Comment:
Instead of copying the whole file, isn't there something exposed in that packages (or other packages of standard library) that just allows the opening of a dialed connection through a target SOCKS server?